### PR TITLE
Implement GetZoneByProviderID and GetZoneByNodeName for openstack

### PR DESCRIPTION
**Pull request by FengyunPan**
This is part of #50926
cc @wlan0

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 1]
```diff
@@ -549,18 +553,66 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
 	return zone, nil
 }
 
+func (os *OpenStack) getZoneByServer(server *servers.Server) (cloudprovider.Zone, error) {
+	az := ""
+	var ok bool
+
+	if az, ok = server.Metadata[AvailabilityZone]; !ok {
+		return cloudprovider.Zone{}, fmt.Errorf("Can not get %s from the Metadata of instance %s", AvailabilityZone, server.Name)
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: `nil` should be ok for the return value right? (Same below)

Comment (dims) [reply pkg/cloudprovider/providers/openstack/openstack.go]: for the first parameter

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 2]
```diff
@@ -549,18 +553,66 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
...
+	}
+	if len(az) == 0 {
+		return cloudprovider.Zone{}, fmt.Errorf("The %s of instance %s is empty.", AvailabilityZone, server.Name)
+	}
+
+	zone := cloudprovider.Zone{
+		FailureDomain: az,
+		Region:        os.region,
+	}
+	glog.V(1).Infof("Current zone is %v", zone)
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: Does this need to be `V(1)`? Can we make it `V(4)`

**Comment by luxas** [comment]
PTAL @kubernetes/sig-openstack-misc

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 1 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack.go]: Sorry, nil should be ok, If az is empty, just let caller handle it. Do not return err like GetZone().

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 2 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack.go]: Right, I update it soon.

**Comment by luxas** [comment]
@dims could you please re-review and give your LGTM?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 3]
```diff
@@ -553,14 +557,54 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
 // This is particularly useful in external cloud providers where the kubelet
 // does not initialize node data.
 func (os *OpenStack) GetZoneByProviderID(providerID string) (cloudprovider.Zone, error) {
-	return cloudprovider.Zone{}, errors.New("GetZoneByProviderID not implemented")
+	instanceID, err := instanceIDFromProviderID(providerID)
+	if err != nil {
+		return cloudprovider.Zone{}, err
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: can we return `nil, err`?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 4]
```diff
@@ -553,14 +557,54 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
...
+	}
+
+	compute, err := os.NewComputeV2()
+	if err != nil {
+		return cloudprovider.Zone{}, err
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: same here

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 5]
```diff
@@ -553,14 +557,54 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
...
+	}
+
+	srv, err := servers.Get(compute, instanceID).Extract()
+	if err != nil {
+		return cloudprovider.Zone{}, err
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: same here

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 6]
```diff
@@ -553,14 +557,54 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
...
 }
 
 // GetZoneByNodeName implements Zones.GetZoneByNodeName
 // This is particularly useful in external cloud providers where the kubelet
 // does not initialize node data.
 func (os *OpenStack) GetZoneByNodeName(nodeName types.NodeName) (cloudprovider.Zone, error) {
-	return cloudprovider.Zone{}, errors.New("GetZoneByNodeName not imeplemented")
+	compute, err := os.NewComputeV2()
+	if err != nil {
+		return cloudprovider.Zone{}, err
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: same here

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 7]
```diff
@@ -553,14 +557,54 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
...
+	}
+
+	srv, err := getServerByName(compute, nodeName)
+	if err != nil {
+		if err == ErrNotFound {
+			return cloudprovider.Zone{}, cloudprovider.InstanceNotFound
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: same here

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 3 above)
Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack.go]: No, nil is incorrect, we cannot use nil as type cloudprovider.Zone in return argument. The "cloudprovider.Zone{}" is ok.

Comment (FengyunPan) [reply pkg/cloudprovider/providers/openstack/openstack.go]: The return argument is not a point type, it is a object, nil is incorrect.

**Comment by dims** [comment]
Ack thanks @FengyunPan 

/approve

**Comment by FengyunPan** [comment]
Thank for your labels.

**Pull request merged**